### ISSUE TITLE
Cannot process the same action from 2 different effects

### ISSUE THREAD

**Issue by DxCx** [ngrx/effects#87]
Hi,

I'm trying to use Dispatcher.INIT action, to do some effects init,
i have couple of modules that should process it, but looks like it is trapped.
(Only 1 module will process it, if i remove it from NgModule, only the next will process, etc..)

so it looks something like this:
app.module.ts
```typescript
@NgModule({
    ....
    imports: [
      EffectsModule.run(ModuleAEffects),
      EffectsModule.run(ModuleBEffects),
    ],
    ....
})
export class AppModule {};
```

then each module Effects class will have a simple pipe of:
module<x>.effects.ts
```typescript
@Effect() init$ =
this.actions$.ofType(Dispatcher.INIT)
  .switchMap(() => {
      // DO INIT FOR MODULE
  });
```

and as i mentioned above, only the first one registered on NgModule is being processed,
the other does not.

then, i said to myself, let's try to split the 1 trapped init event, into 2 events (or more in the future)
so i resulted creating actions of `MODULE_A_INIT`, `MODULE_B_INIT`.
replace the `ofType` above for each module,
then created another effect which is registered first and containing:

```typescript
@Effect() init$ =
this.actions$.ofType(Dispatcher.INIT)
  .switchMap(() => Observable.of([{
    type: MODULE_A_INIT,
  }, {
    type: MODULE_B_INIT,
  }])
  .concatMap(actions => Observable.from(actions)));
```
NOTE: i tried also with `mergeMap` and `flatMap`

and with this approach, i can see both actions are fired, however the first _INIT is not being processed, only the lastest action emitted is. (switchMap/withLatestFrom?)

so, TLDR, how can i process the same action using more then 1 effect?
for reducers it works, you can process the same action in more then one reducer..

**Comment by DxCx** [comment]
Nothing... :(
@kamilkisiela @urigo @dotansimha maybe one of you encountered this issue?

**Comment by kayue** [comment]
Having the same issue.

**Comment by bertofer** [comment]
It is working for me, I have an action that just logs any action that goes through the store, and that doesn't prevent me for creating side-effects on another Effects module to actually process the action.

Here is my logger effect:
https://github.com/bertofer/arxivum-web/blob/develop/src/app/app.effects.ts

And here authentication effects:
https://github.com/bertofer/arxivum-web/blob/develop/src/app/core/authentication/authentication.effects.ts

I am adding the import module as you in the app.module.ts, so I don't know what may be happening in your case, can you provide more info? ngrx/effects version, repo if the code is open source...

**Comment by DxCx** [comment]
interesting, from what i see in your example, the logger does not catch the events,
but just let them flow,
maybe that's my issue, i was afraid that i i'll generate the same event again, it  will run recursively.
i'll try it out and let you know

**Comment by DxCx** [comment]
Ok, so after checking this up:
1. I need to use switchMap as i want to catch errors and make them into something useful,
when using switchMap and returning the same original action as a new Observable.of() it will call it recursively.
2. as part of my INIT, some of the effects needs to generate more events, i'm not able to create an event and populate to current one. so for example, i want to do login on init if there is a key inside the localStorage.
then on INIT i'll check localStorage, and if key exists i want both the INIT to continue and a new event to be generated with LOGIN as type and key as payload.

if i'll create a new example repo showing this issue, will it help?

**Comment by bertofer** [comment]
Yes, will be useful as we can try it locally.

Regarding the recursivity, it can happen. If you don't use @Effect({dispatch: false}) on the logger for instance, it will propagate @ngrx/store/init and it will be dispatched again entering an infinite loop. 

What do you mean by `the logger does not catch the events`? Catch how?

Regarding 1., for now I am using store.dispatch() in the .catch error, because I think ngrx/effects doesn't dispatch when the action comes directly from the catch, I am not sure if this can be a rxjs limitation. I want to investigate more this to open an issue if needed.

In 2., what do you mean by events? Actions? Side effects?

**Comment by DxCx** [comment]
hi @bertofer thanks for the willing to help!

so for the recursiveness, yes, i missed the effect flag, once applying dispatch: false it is indeed not recursive.

now, for my comment about `the logger does not catch the events`, 
what i meant is that you are not trapping.
which means, you are processing it, and then send it again as a value instead of returning another action instead or just Observable.empty().

so, thanks to you, i've learned that using `@Effect({dispatch: false})` 
gives the ability to listen to actions without trapping them.
however, if i want the @Effect to dispatch 1 or more actions as a result, it would not work.

to solve this, i've found out that store.dispatch can be called, however, it feels abit intuitive to me,
so the last result looks like:
```typescript
// constructors set private store: Store
@Effect({ dispatch: false }) init$ =
this.actions$.ofType(Dispatcher.INIT)
  .switchMap(() => Observable.of([{
    type: MODULE_A_INIT,
  }, {
    type: MODULE_B_INIT,
  }])
  .concatMap(actions => Observable.from(actions)))
  .do((action) => this.store.dispatch(action));
```

can you suggest a better solution then this?
also, (Maybe that's not to you but to @Effect author) why not supporting list of actions, so if an @Effect spits out a list of actions, they will be dispatched in-order?

**Comment by bertofer** [comment]
Yes, I agree that would be a great feature to be able to return an Array of actions and dispatch them all.

For your workaround, are you sure you need to switchMap to 2 different actions to treat them differently, instead of doing the side-effects in the INIT itself?

If you need to trigger different actions, you can use Observable.merge of different async operations , something like: 

```javascript
 asyncOp = (action) => {
    return Observable.of(action).delay(100).map(action => {
      return {
        type: 'ACTION',
        payload: action
      };
    }).catch(err => Observable.of(err));
  }

@Effect() init$ = this.actions$
    .ofType(Dispatcher.INIT)
    .switchMap(action => {
      return Observable.merge(
        this.asyncOp(action),
        this.asyncOp(action)
      );
    })
```

This works for me, dispatching 2 (potentially) different actions that are registered in the logger:

(first is type of action, then payload)

<img width="492" alt="screen shot 2017-01-22 at 19 21 37" src="https://cloud.githubusercontent.com/assets/1634229/22185144/19dff146-e0d8-11e6-9eeb-d073f48613f1.png">

Changing the asyncOp function by your http requests should work.

**Comment by kayue** [comment]
Hi 

I have created a Plunker to reproduce the original issue; only the first effect is being called. 

https://plnkr.co/edit/OIYdo8Mm4VzwPjdyiEXB?p=preview

Looks like the problem is with `switchMap`, once the first effect is called, the `@ngrx/store/init` will not pass to the next effect. Therefore it only being called once.

`mergeMap` can be used to dispatch multiple action, but it will cause infinite loop. (And setting `dispatch: false` doesn't work)

```
this.actions$.ofType('@ngrx/store/init').mergeMap(action => {
  return Observable.of(action, { type: 'SECOND_ACTION', payload: [] });
}
```

**Comment by DxCx** [comment]
hi @kayue out of cariousity i played with your plunkr,
and found out that this works there for some reason:
```
  @Effect()
  init$ = this.actions$.ofType('@ngrx/store/init').mergeMap(action => {
    return Observable.of({type: 'FIRST_ACTION', payload: []}, 
                         { type: 'SECOND_ACTION', payload: [] });
  })
}
```

**Comment by kayue** [comment]
@DxCx Not exactly. I have another Plunkr here to reproduce the problem

https://plnkr.co/edit/kHjHQ4Jx2RBv8kDpGMtn?p=info

Although it did dispatch the `FIRST_ACTION`, but somehow the message `1st action success` is not showing... :/

**Comment by bertofer** [comment]
I have an idea why this might be happening. 

So, first of all, the dispatcher that dispatches actions is a BehaviorSubject, which means you receive the last value emitted when you are subscribed to it. 

And the initialisation of the disopatcher, where it sets the first value, is called **before** the subscriptions of the Effects. So, this is what I think it's happening:

1. Dispatcher.INIT action is dispatched.
2. Then this code is executed:
```javascript
@Effect()
  init$ = this.actions$.ofType('@ngrx/store/init')
  .switchMap(action => {
    return Observable.of({ type: 'FIRST_ACTION', payload: [] })
  });
```
this code is **synchronously** pushing a  new Action to the dispatcher, which means that when this code is executed:

```javascript
@Effect()
  init$ = this.actions$.ofType('@ngrx/store/init')
  .switchMap(action => {
    return Observable.of({ type: 'SECOND_ACTION', payload: [] });
  });
```
The **last value** of the dispatcher (from where the actions$ stream is created) is already the FIRST_ACTION. This only happens at the beginning, because the `@Effects` where not yet subscribed to it, while after that, as they are already subscribed, will receive all the values.

For instance, in the example that @kayue posted, both FIRST_ACTION and SECOND_ACTION are synchronously added to the dispatcher before the `@Effect` is subscribed, so when the effects are subscribed, the last action was SECOND_ACTION only.

Turns out, by adding `delay(1)` to the dispatch, making it async and at the bottom of the event cycle of javascript, the `@Effects` can subscribe, and then everything works fine.

```javascript
// In this case works
@Effect()
  init$ = this.actions$.ofType('@ngrx/store/init')
  .mergeMap(action => {
    return Observable.of(
      { type: 'FIRST_ACTION', payload: [] },
      { type: 'SECOND_ACTION', payload: [] }
    )
  }).delay(1);
```

From here, I am not sure is this should be considered an issue itself, and if it's the case, what repository should it be (here, or @ngrx/store ? )

Would appreciate if anyone could confirm/deny/develop on this explanation.

**Comment by DxCx** [comment]
@bertofer that was exactly my tought, and then i saw that it works without delay as well.
but maybe i missed it and one of the actions was just showen and not actually proccessed..

**Comment by bertofer** [comment]
@DxCx Can you show me how you made it work without delay?

Actually there are a lot of ways to make it work, you only need to make an async op so that the effects can subscribe, and from there on they can listen to all actions

**Comment by DxCx** [comment]
@bertofer, @kayue already pointed out that i missed it, but it didnt fully work.
And assuming this is not working then yeah im totally with you on the theory above.

Regarding the fix, sounds like this is the right behavior for BehaviorSubject so IMO what we really need is a patch for effects that if array is emitted it will break it to async events...

**Comment by bertofer** [comment]
I am thinking that maybe the Dispatcher.INIT action it's just not the right action to capture when you want to initialize your state, as it's a function specially there to initialize the reducers with the initial state. I think one approach would be to just dispatch a custom INIT action when the AppComponent is created, and do all the initialisation there.

**Comment by DxCx** [comment]
@bertofer, init is just a common case, i still think a user might want to dispatch the same action or 2 actions out of one, and not necessarily INIT

**Comment by bertofer** [comment]
@DxCx I think it works fine in that case, the only problem is with the Dispatcher.INIT action, because the effects aren't yet subscribed to the dispatcher when this action happens (It is actually dispatched in the constructor of the dispatcher in @ngrx/store), but for all the following actions, all the effects should work as all of them are already subscribed to actions, so will receive any new action dispatched.

**Comment by bryanrideshark** [comment]
To solve the issue brought up by @bertofer , could there be an Effects specific INIT action which is dispatched when all of the effects are subscribed and ready to rumble?

**Comment by robwormald** [comment]
Please check this against NgRx v4, and if it’s still an issue, please reopen on https://github.com/ngrx/platform. Thanks!
